Output largo specific photo credit metadata on image block

joshdarby opened the pull request:
## Changes

This pull request makes the following changes:

- Uses the new blocks API to insert media credits into images in the image block.

![Screen Shot 2019-07-03 at 10 51 53 AM](https://user-images.githubusercontent.com/18353636/60601730-943e5b00-9d80-11e9-9e9a-fc114b8b1323.png)

## Why

<!-- Why does this PR propose these changes? Take as much space as you need to explain. -->
<!-- If there are GitHub issues that this pull request addresses, please list them here. -->
For #1683.

## Testing/Questions

Features that this PR affects:

- Image blocks in Gutenberg.

<!-- If there are no questions, please remove the questions section. -->
Questions that need to be answered before merging:

- [ ] What could we change to make this better?
- [ ] Is the current placement/styling of the media credit ok?

Steps to test this PR:

1. Edit Gutenberg post
2. Insert 2 image blocks into post
   a. One image with a media credit with no url
   b. One image with a media credit with a url
3. Once inserted, save the post , refresh/click around and then preview it.
4. Does the media credit show up correctly on the frontend?

Commit:
Rough working version of outputting media credit in block editor

Commit:
Fix gross spacing issues

Commit:
Moved gutenberg block edit functions to inc/ file

Commit:
Added _media_credit_url meta field to REST endpoint, added it to caption

Commit:
Added more checks before adding in our media credit/url

Commit:
Make sure the entire media_credit obj exists before checking contents

Commit:
Removing all of the unused constants

Commit:
Updated function name to be more relevant

Commit:
Updated changelog.md

Commit:
Merge branch '0.5-dev' into 1683-output-largo-speicifc-photo-credit-metadata-on-image-block

benlk commented on inc/gutenberg-block-edits.php:
```diff
@@ -0,0 +1,61 @@
+	);
+
+}
+add_action( 'rest_api_init', 'largo_register_custom_rest_fields' );
```
This should probably go in the file where we register that metadata to begin with.

benlk commented:
Questions for review:
- [ ] does it work in:
    - [ ] current wp core
    - [ ] current gutenberg plugin
- [ ] could it be improved in any way that's achievable with little time input?
- [ ] does it have the possibility of messing with anything?

benlk commented:
- [ ] Does Gutenberg content that already exists automatically have this added in the output, or will posts have to be resaved?
    - [ ] Should it? This is something that we'd want to poll users on.

benlk commented:
An image inserted before this change, using WP 5.2.2 and with the Gutenberg plugin inactive, results in block validation failure for `core/image` blocks:

> Content generated by `save` function:
> 
> <figure><img src="https://largo.test/wp-content/uploads/2019/07/prisons-vs-intertidal-1.png" alt="A map of Louisiana overlaid with a shadow that brings the shorelines of Louisiana well inland. The boot is cut in twain."/><figcaption>The shadow shows the boundaries of Louisiana if intertidal lands and wetlands are not counted towards the state's territory. State correctional facilities are marked.</figcaption></figure>
> 
> Content retrieved from post body:
> 
> <figure class="wp-block-image"><img src="https://largo.test/wp-content/uploads/2019/07/prisons-vs-intertidal-1.png" alt="A map of Louisiana overlaid with a shadow that brings the shorelines of Louisiana well inland. The boot is cut in twain." class="wp-image-629"/><figcaption>The shadow shows the boundaries of Louisiana if intertidal lands and wetlands are not counted towards the state's territory. State correctional facilities are marked.</figcaption></figure>

```diff
- <figure                       ><img src="https://largo.test/wp-content/uploads/2019/07/prisons-vs-intertidal-1.png" alt="A map of Louisiana overlaid with a shadow that brings the shorelines of Louisiana well inland. The boot is cut in twain."                     /><figcaption>The shadow shows the boundaries of Louisiana if intertidal lands and wetlands are not counted towards the state's territory. State correctional facilities are marked.</figcaption></figure>
+ <figure class="wp-block-image"><img src="https://largo.test/wp-content/uploads/2019/07/prisons-vs-intertidal-1.png" alt="A map of Louisiana overlaid with a shadow that brings the shorelines of Louisiana well inland. The boot is cut in twain." class="wp-image-629"/><figcaption>The shadow shows the boundaries of Louisiana if intertidal lands and wetlands are not counted towards the state's territory. State correctional facilities are marked.</figcaption></figure>
```

It looks like the change introduced in this PR is failing to save the `.wp-block-image` class on the containing `<figure>` element, and the image ID class on the `<img>`.

benlk commented:
On the first save of a post containing a newly-inserted image block for an image that contains a media credit, the `post_content` does not contain the image credit. After that save, the image credit appears in the editor, but until the next post save, the editor shows something that doesn't exist in the post_content or on the front end.

The media credit doesn't capture the organization field from the media credit metadata, in my testing.

benlk commented on inc/gutenberg-block-edits.php:
```diff
@@ -0,0 +1,61 @@
+
+	$meta = get_post_meta( $post_id );
+	
+    $meta_fields = [ '_media_credit', '_media_credit_url' ];
```
Should also include _navis_media_credit_org and _navis_media_can_distribute

I'm not sure what _navis_media_can_distribute does, but I think it should be in the API if we're including all the other info.

benlk commented on js/blocks-image-customization.js:
```diff
@@ -0,0 +1,53 @@
+            if( media.media_credit && media.media_credit._media_credit ){
+
+                // our default media credit, if it exists 
+                var media_credit = '<span class="largo-attachment-media-credit">' + media.media_credit._media_credit + '</span>';
```
The existing markup for this is:

```
<p class="wp-media-credit">credit / organization</p>
```

The generator for that markup is https://github.com/INN/largo/blob/0.5-dev/lib/navis-media-credit/navis-media-credit.php#L242-L273

benlk commented on js/blocks-image-customization.js:
```diff
@@ -0,0 +1,53 @@
+                // else, our media credit probably doesn't exist. let's add it in
+                } else {
+
+                    var attachment_caption = attachment_caption + '<br/>' + media_credit;
```
To match existing caption markup, the media credit should be prepended. 


```html
<div id="attachment_11" class="wp-caption module image alignnone" style="max-width: 336px;">
    <img class="size-medium wp-image-11" src="https://largo.test/wp-content/uploads/2018/05/NeilPatrickHarris-moustache-cr-336x199.jpeg" alt="alt text alt text alt text" srcset="https://largo.test/wp-content/uploads/2018/05/NeilPatrickHarris-moustache-cr-336x199.jpeg 336w, https://largo.test/wp-content/uploads/2018/05/NeilPatrickHarris-moustache-cr.jpeg 633w" sizes="(max-width: 336px) 100vw, 336px" width="336" height="199">
    <p class="wp-media-credit">credit credit / organization</p>
    <p class="wp-caption-text">Caption caption caption</p>
</div>
```

benlk commented:
But all that criticism aside: This is a massive improvement over the status quo, and it proves that the thing we're trying to do is, in theory, possible. 🎉

joshdarby commented:
@benlk Do you think it's worth hacking our way through this right now instead of using something like `DOMDocument` to render the caption in the post content in the interim while we wait for a future Gutenberg release to address this?

benlk commented:
If there's no way to make it work on the [`blocks.getSaveElement` hook/filter](https://developer.wordpress.org/block-editor/developers/filters/block-filters/#blocks-getsaveelement) , then yes, I think it would be better to do the serverside frontend-only approach. There's a number of filters within the function [render_block](https://developer.wordpress.org/reference/functions/render_block/) that might allow us to generate appropriate markup without using DOMDocument or [regexes in HTML](https://stackoverflow.com/a/1732454).

benlk commented:
Coming in Gutenberg 6.1: https://developer.wordpress.org/block-editor/components/slot-fill/

Commit:
Merge branch '0.5-dev' into 1683-output-largo-speicifc-photo-credit-metadata-on-image-block

joshdarby commented:
@benlk Looking more into this here https://github.com/WordPress/gutenberg/issues/10204 brings up a good point. If we continue down the path of manipulating the image block and then someone chooses to switch off of Largo, all of their image blocks (presumably) will go into the "Invalid" state since it won't match its expected output.

benlk commented:
I'm not sure that we need to be worried about that, since we're not adding content that is invalid in the area that we're adding it.

The examples in https://github.com/WordPress/gutenberg/issues/10204 show adding new `data-` attributes to elements. We're adding some `<span>` tags to the contents of the `<figcaption>` element.

I'm currently running the #1662 branch of Largo on a database where I had previously used this branch to test the addition of elements. This gives us the setup: where the code in this PR had been active, but is not.

Here's the big conflict:

![Screen Shot 2019-07-15 at 12 43 49 ](https://user-images.githubusercontent.com/1754187/61233381-736df200-a6fe-11e9-87ba-e74903f4054b.png)

Parts:
- Gutenberg thinks there should be a `wp-block-image` class on the `<figure>`, which was an issue with this PR that I raised in https://github.com/INN/largo/pull/1733#issuecomment-510693102
- Gutenberg doesn't like the `/` in `<br/>` because that's XHTML, not HTML, but it's not a big worry. https://stackoverflow.com/questions/1946426/html-5-is-it-br-br-or-br

joshdarby commented:
Thanks for looking at that so quickly.

Commit:
Move registration of media credit data for api to relevant file, prepend

media credit, and also use correct class and structure

Commit:
Switch from using blocks.getSaveContent to editor.BlockEdit

joshdarby commented:
@benlk I think this is ready for review again. I addressed all of your comments. I also switched from using `blocks.getSaveContent` to `editor.BlockEdit` so the credit gets added in real time.

It also seems to be backwards compatible now and doesn't break blocks when you switch to a different branch without this code.

Commit:
Removed unused `_extends` function

benlk commented on js/blocks-image-customization.js:
```diff
@@ -0,0 +1,55 @@
+                // our default media credit, if it exists 
+                var media_credit = '<p class="wp-media-credit">' + media.media_credit._media_credit + '</p>';
+
+                // if our media credit has a url, include that also
```
This section tests if it has a media credit org, not if it has a URL.

This PR does not, presently, include a function for including the URL in the media credit.

joshdarby replied:
@benlk I wasn't sure about the URL since it was set as a variable but wasn't used in the classic editor way here:
https://github.com/INN/largo/blob/39a0fc4ca34a118ae6ae870dd591029f5a61fdb1/lib/navis-media-credit/navis-media-credit.php#L248-L250

Should the URL wrap the credit or organization name?

benlk replied:
Odd that it isn't used in the main media credit instance.

It's used for the "hero" featured image on posts, where the link wraps the credit and org name together: https://github.com/INN/largo/blob/72cc8433a1be1de4f22789e1ec89496d54a653d9/partials/hero-featured-image.php#L5-L7

benlk reviewed (changes requested):
Needs support for the media credit URL, but apart from that I think this is good to go. 👏

benlk commented:
This PR doesn't modify the Gallery Block, but if we want to cover that I think it should be in a separate PR, for a separate issue.

Commit:
Wrap our media credit in the media credit url if it exists

Commit:
Merge branch '0.5-dev' into 1683-output-largo-speicifc-photo-credit-metadata-on-image-block

joshdarby commented:
@benlk https://github.com/INN/largo/pull/1733/commits/0660b386bf6a45b43e97266de4670f2f9b6d295b updates the media credit to include the url if it exists.

So our test cases should be an image with:

- [ ] Credit + Org + URL
- [ ] Credit + URL
- [ ] Credit + Org
- [ ] Credit

benlk commented:
has/doesn't of:

- credit
- org
- URL

But the case where an image has a URL alone makes little sense. So let's add:

- [ ] org + URL
- [ ] org alone
- [ ] no credit metadata at all

joshdarby commented:
@benlk I'll have to rewrite part of this if we want these two to work:

- org + URL
- org alone

Right now the first conditional just checks if the credit exists and doesn't show anything if the credit doesn't exist.
```
if( media.media_credit && media.media_credit._media_credit ){
```

Commit:
Allow organization to display alone if the credit doesn't exist

joshdarby commented:
https://github.com/INN/largo/pull/1733/commits/5ffccafabdf4ea320a7972980440565942b50979 updates the conditional to allow the organization (+url if it exists) to display if the credit doesn't exist.

benlk commented:
If that worked in your testing, let's merge it and be done.
